src/rt/runtime.rs

```diff
@@ -0,0 +1,449 @@
+
+    /// Schedules a task.
+    pub fn schedule(&self, task: Runnable) {
+        MACHINE.with(|machine| {
+            // If the current thread is a worker thread, schedule it onto the current machine.
+            // Otherwise, push it into the global task queue.
+            match machine.get() {
+                None => {
+                    self.injector.push(task);
+                    self.notify();
+                }
+                Some(m) => m.schedule(&self, task),
+            }
+        });
+    }
+
+    /// Runs the runtime on the current thread.
+    pub fn run(&self) {
+        scope(|s| {
+            let mut idle = 0;
+            let mut delay = 0;
+
+            loop {
+                // Get a list of new machines to start, if any need to be started.
+                for m in self.make_machines() {
+                    idle = 0;
+
+                    s.builder()
+                        .name("async-std/machine".to_string())
+                        .spawn(move |_| {
+                            abort_on_panic(|| {
+                                let _ = MACHINE.with(|machine| machine.set(m.clone()));
+                                m.run(self);
+                            })
+                        })
+                        .expect("cannot start a machine thread");
+                }
+
+                // Sleep for a bit longer if the scheduler state hasn't changed in a while.
+                if idle > 10 {
+                    delay = (delay * 2).min(10_000);
+                } else {
+                    idle += 1;
+                    delay = 1000;
+                }
+
+                thread::sleep(Duration::from_micros(delay));
+            }
+        })
+        .unwrap();
+    }
+
+    /// Returns a list of machines that need to be started.
+    fn make_machines(&self) -> Vec<Arc<Machine>> {
+        let mut sched = self.sched.lock().unwrap();
+        let mut to_start = Vec::new();
+
+        // If there is a machine that is stuck on a task and not making any progress, steal its
+        // processor and set up a new machine to take over.
+        for m in &mut sched.machines {
+            if !m.progress.swap(false, Ordering::SeqCst) {
+                let opt_p = m.processor.try_lock().and_then(|mut p| p.take());
+
+                if let Some(p) = opt_p {
+                    *m = Arc::new(Machine::new(p));
+                    to_start.push(m.clone());
+                }
+            }
+        }
+
+        // If no machine has been polling the reactor in a while, that means the runtime is
+        // overloaded with work and we need to start another machine.
+        if !sched.polling {
+            if !sched.progress {
+                if let Some(p) = sched.processors.pop() {
+                    let m = Arc::new(Machine::new(p));
+                    to_start.push(m.clone());
+                    sched.machines.push(m);
+                }
+            }
+
+            sched.progress = false;
```

Comment:
@stjepang 
I have one question.
Why set false to sched.progress here? I'm interested.